### ISSUE TITLE
`func-style` `allowArrowFunctions` option should work with `expression`

### ISSUE THREAD

**Issue by ehmicky** [eslint/eslint#13141]
**What rule do you want to change?**

`func-style`

**Does this change cause the rule to produce more or fewer warnings?**

More warnings.

**How will the change be implemented? (New option, new default behavior, etc.)?**

The option `allowArrowFunctions: false` should work with `expression` not only with `declaration` 

**Please provide some example code that this change will affect:**

```js
const one = () => true

function two() {
  return true
}

const three = function() {
  return true
}
```

**What does the rule currently do for this code?**

When using the following configuration:

```yml
rules:
  func-style:
    - 2 
    - expression
    - allowArrowFunctions: false
```

`two` should print a warning, but `one` and `three` do not.

**What will the rule do after it's changed?**

`one` would now print a warning.

**Comment by kaicataldo** [comment]
This looks like it's working as intended to me, since arrow functions are always expressions. Do you mind clarifying if I'm misunderstanding?

**Comment by ehmicky** [comment]
Oh sorry, I meant `allowArrowFunctions: false` not `allowArrowFunctions: true`. I updated my original post.

**Comment by kaicataldo** [comment]
Can you explain the problem you're trying to solve with this? Since arrow function expressions can't be converted to function declarations, this would effectively disallow variable declarations with arrow function expressions, and I think that's beyond the scope of the rule.

If you would like to enforce this, I think using `no-restricted-syntax` is a better option. You can see an example using our demo [here](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXCJlcnJvclwiLCBcIlZhcmlhYmxlRGVjbGFyYXRvciA+IEFycm93RnVuY3Rpb25FeHByZXNzaW9uXCJdICovXG5cbmNvbnN0IG9uZSA9ICgpID0+IHRydWVcblxuZnVuY3Rpb24gdHdvKCkge1xuICByZXR1cm4gdHJ1ZVxufVxuXG5jb25zdCB0aHJlZSA9IGZ1bmN0aW9uKCkge1xuICByZXR1cm4gdHJ1ZVxufSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6Niwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==).

**Comment by ehmicky** [comment]
I am trying to enforce style consistency. 

```js
const one = () => {
  ...
}

function two() {
  ...
}

const three = function() {
  ...
}
```

There are some differences in those statements, among: 
  - `two` is a declaration, not an expression
  - `two` gets hoisted
  - `one.prototype === undefined` while `three.prototype !== undefined` 

However, for most of the cases, those are just three mostly equivalent ways of declaring a function. Some developers might not want both `three` and `one` mixed up in the same codebase. From that standpoint I think it would go in line with what `func-style` is trying to achieve.

That being said, `no-restricted-syntax` does solve this issue, although in a more indirect way.

**Comment by kaicataldo** [comment]
I still personally feel like this goes beyond the scope of responsibility for this rule, but I'm happy to defer to consensus here. Regarding the autofix, I don't think arrow functions should be autofixed, because it has the potential to change semantics and execution.

**Comment by ehmicky** [comment]
Just to be completely precise: for the autofix, I was referring to `one` being converted to `three`, or the other way around. Not converting `two` to `one` or `three`, since that would definitely bring potential breaking changes.

When it comes to converting `const one = () => {}` from/to `const one = function() {}`, the only behavioral change I can think of is that `one.prototype` is `undefined` in one case but not the other. Is there something else?

That being said again, I can close this issue if this rule addition does not makes sense to you.
